workflow/controller/operator.go

```diff
@@ -1768,6 +1773,15 @@ func (woc *wfOperationCtx) executeTemplate(nodeName string, orgTmpl wfv1.Templat
 	return node, nil
 }
 
+func (woc *wfOperationCtx) appendFailHostsToAffinity(tmpl *wfv1.Template, retryNodeName string) *wfv1.Template {
```

Comment:
* move to own file
* this need not/should not be a receiver func (i.e. remove `woc`)
* remove return value - you already modify the param so this is redundant

create a new interface, something like

```golang
type RetryTweak = func (retryStrategy RetryStrategy, previousAttempts wfv1.Nodes, tmpl *wfv1.Template)
```

this func can then implement that interface

Reply:
I implemented this in a new file.